File: bullmq_worker/storage/aliyun_oss_client.py
import os
import oss2
from typing import  Tuple
from urllib.parse import unquote, urlparse

from bullmq_worker.utils.logging import logger
from bullmq_worker.storage.base_storage_client import BaseStorageClient


class AliyunOSSClient(BaseStorageClient):
    """阿里云OSS客户端"""

    def __init__(self, endpoint: str, access_key: str, secret_key: str, region: str, bucket_name: str):
        super().__init__(endpoint, access_key, secret_key, region, bucket_name)

        # 创建认证对象
        auth = oss2.Auth(self.access_key, self.secret_key)

        # 创建Bucket对象
        self.bucket = oss2.Bucket(auth, self.endpoint, self.bucket_name)

        # 检查桶是否存在
        self._ensure_bucket_exists()

    def _ensure_bucket_exists(self):
        """阿里云OSS的桶需要在控制台创建，这里只检查是否存在"""
        try:
            self.bucket.get_bucket_info()
            logger.info(f"桶 {self.bucket_name} 存在")
        except oss2.exceptions.NoSuchBucket:
            logger.error(f"桶 {self.bucket_name} 不存在")
            raise RuntimeError(f"桶 {self.bucket_name} 不存在")
        except Exception as e:
            logger.error(f"检查桶状态失败: {e}")
            raise RuntimeError(f"检查桶状态失败: {e}")

    def download_file(self, object_key: str, local_path: str) -> str:
        """从阿里云OSS下载文件"""
        try:
            # 创建本地目录
            os.makedirs(os.path.dirname(local_path), exist_ok=True)

            # 下载文件
            self.bucket.get_object_to_file(object_key, local_path)

            logger.info(f"下载成功: {self.bucket_name}/{object_key} -> {local_path}")

        except oss2.exceptions.NoSuchKey:
            logger.error(f"文件不存在: {self.bucket_name}/{object_key}")
            raise RuntimeError(f"文件不存在: {self.bucket_name}/{object_key}")
        except oss2.exceptions.OssError as e:
            logger.error(f"阿里云OSS下载失败: {e}")
            raise RuntimeError(f"阿里云OSS下载失败: {e}")

    def upload_file(self, object_key: str, local_path: str) -> str:
        """上传文件到阿里云OSS"""
        try:
            # 检查本地文件是否存在
            if not os.path.exists(local_path):
                raise FileNotFoundError(f"本地文件不存在: {local_path}")
            
            # 上传文件
            with open(local_path, "rb") as file_obj:
                self.bucket.put_object(object_key, file_obj)
            logger.info(f"上传成功: {local_path} -> {self.bucket_name}/{object_key}")
            
            # 生成预签名URL
            output_url = self.build_url(object_key)

            return output_url

        except oss2.exceptions.OssError as e:
            logger.error(f"阿里云OSS上传失败: {e}")
            raise RuntimeError(f"阿里云OSS上传失败: {e}")
    
    def build_url(self, object_key: str, signer_url_expire_hours: int = 24) -> str:
        """构建阿里云OSS URL - 生成预签名URL以支持私有文件访问"""
        try:
            expire_seconds = signer_url_expire_hours * 3600
            signed_url = self.bucket.sign_url("GET", object_key, expire_seconds)

            logger.info(
                f"生成预签名URL: {signed_url}，有效期: {signer_url_expire_hours}小时"
            )
            return signed_url
        
        except Exception as e:
            logger.error(f"生成预签名URL失败: {e}")
            raise RuntimeError(f"生成预签名URL失败: {e}")

    def parse_url(self, url: str) -> Tuple[str, str]:
        """解析阿里云OSS URL"""
        parsed = urlparse(url)
        hostname = parsed.hostname or ""

        # 阿里云OSS URL格式: https://bucket.oss-region.aliyuncs.com/path/file
        if ".aliyuncs.com" in hostname:
            parts = hostname.split(".")
            if len(parts) >= 3:
                bucket = parts[0]
                object_key = unquote(parsed.path.lstrip("/"))
                return bucket, object_key

        raise ValueError(f"无法解析阿里云OSS URL: {url}")


def test_client():
    from dotenv import load_dotenv, find_dotenv
    load_dotenv(find_dotenv())

    client = AliyunOSSClient(
        endpoint=os.getenv("S3_ENDPOINT"),
        access_key=os.getenv("S3_ACCESS_KEY"),
        secret_key=os.getenv("S3_SECRET_KEY"),
        bucket_name=os.getenv("S3_BUCKET"),
        region=os.getenv("S3_REGION")
    )

    # 上传文件
    local_path = "assets/demo1_audio.wav"
    url = client.upload_file(
        object_key="wjw_demo1_audio.wav",
        local_path=local_path
    )

    # 下载文件
    bucket, object_key = client.parse_url(url)
    client.download_file(
        object_key=object_key,
        local_path=f"downloads/{object_key}"
    )

if __name__ == "__main__":
    test_client()

File: bullmq_worker/src/storage/aliyun-oss-client.ts
import OSS from 'ali-oss';
import * as fs from 'fs';
import * as path from 'path';
import { URL } from 'url';
import { BaseStorageClient, StorageConfig } from './base-client';
import logger from '../utils/logger';

/**
 * Aliyun OSS storage client implementation
 */
export class AliyunOSSClient extends BaseStorageClient {
  private client: OSS;
  
  // 上传优化配置（可根据网络环境调整）
  private readonly MULTIPART_THRESHOLD = 10 * 1024 * 1024; // 10MB，超过此大小启用分片
  private readonly PART_SIZE = 10 * 1024 * 1024; // 10MB，每个分片大小（建议5-20MB）
  private readonly MAX_CONCURRENT_PARTS = 3; // 最大并行分片数（建议2-5，网络好可增加到8）
  
  constructor(config: StorageConfig) {
    super(config);
    
    // Create Aliyun OSS client
    this.client = new OSS({
      region: this.region,
      accessKeyId: this.accessKey,
      accessKeySecret: this.secretKey,
      bucket: this.bucketName,
      endpoint: this.endpoint,
      secure: this.endpoint.includes('https'),
      // 增加超时配置，解决大文件上传超时问题
      timeout: 1800000, // 30分钟超时时间
      requestTimeout: 1800000, // 请求超时时间
      responseTimeout: 1800000, // 响应超时时间
    });
    
    // 检查桶是否存在
    this.ensureBucketExists().catch(error => {
      logger.error(error.message || error);
    });
  }
  
  /**
   * 确保存储桶存在 (阿里云OSS的桶需要在控制台创建)
   */
  protected async ensureBucketExists(): Promise<void> {
    try {
      await this.client.getBucketInfo(this.bucketName);
      logger.info(`📦 桶 ${this.bucketName} 存在`);
    } catch (error: any) {
      if (error.code === 'NoSuchBucket') {
        throw new Error(`桶 ${this.bucketName} 不存在`);
      } else {
        throw new Error(`检查桶状态失败: ${error}`);
      }
    }
  }
  
  /**
   * 从阿里云OSS下载文件
   */
  async downloadFile(objectKey: string, localPath: string): Promise<void> {
    try {
      // 创建本地目录
      const dir = path.dirname(localPath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      
      // 下载文件
      const result = await this.client.get(objectKey);
      
      // 写入本地文件
      if (result.content) {
        fs.writeFileSync(localPath, result.content);
        logger.info(`✅ 下载成功: ${this.bucketName}/${objectKey} -> ${localPath}`);
      } else {
        throw new Error('❌ 未收到OSS内容');
      }
    } catch (error: any) {
      if (error.code === 'NoSuchKey') {
        logger.error(`❌ 文件不存在: ${this.bucketName}/${objectKey}`);
        throw new Error(`文件不存在: ${this.bucketName}/${objectKey}`);
      } else {
        logger.error(`❌ 阿里云OSS下载失败: ${error}`);
        throw new Error(`阿里云OSS下载失败: ${error}`);
      }
    }
  }
  
  /**
   * 上传文件到阿里云OSS（带重试机制）
   */
  async uploadFile(objectKey: string, localPath: string, maxRetries: number = 3): Promise<string> {
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        // 检查本地文件是否存在
        if (!fs.existsSync(localPath)) {
          throw new Error(`❌ 本地文件不存在: ${localPath}`);
        }
        
        // 获取文件大小
        const stats = fs.statSync(localPath);
        const fileSizeInMB = stats.size / (1024 * 1024);
        
        logger.info(`📤 开始上传文件 (第${attempt}/${maxRetries}次尝试): ${localPath}, 大小: ${fileSizeInMB.toFixed(2)}MB`);
        
        // 对于超过阈值的文件使用分片上传
        if (stats.size > this.MULTIPART_THRESHOLD) {
          logger.info(`📤 使用分片上传处理大文件: ${fileSizeInMB.toFixed(2)}MB (阈值: ${this.MULTIPART_THRESHOLD / (1024 * 1024)}MB)`);
          await this.multipartUpload(objectKey, localPath, stats.size);
        } else {
          // 小文件使用普通上传
          await this.client.put(objectKey, localPath);
        }
        
        logger.info(`✅ 上传成功: ${localPath} -> ${this.bucketName}/${objectKey}`);
        
        // 生成预签名URL
        const outputUrl = await this.buildUrl(objectKey);
        return outputUrl;
        
      } catch (error: any) {
        const isLastAttempt = attempt === maxRetries;
        const errorMessage = error?.message || error?.toString() || '未知错误';
        
        if (isLastAttempt) {
          logger.error(`❌ Aliyun OSS 上传失败 (${maxRetries}次重试后): ${errorMessage}`);
          throw new Error(`Aliyun OSS 上传失败: ${errorMessage}`);
        } else {
          const waitTime = Math.min(1000 * Math.pow(2, attempt - 1), 30000); // 指数退避，最大30秒
          logger.warn(`⚠️ 上传失败，${waitTime/1000}秒后重试 (第${attempt}/${maxRetries}次): ${errorMessage}`);
          
          // 等待后重试
          await new Promise(resolve => setTimeout(resolve, waitTime));
        }
      }
    }
    
    throw new Error('上传失败：所有重试都已用完');
  }
  
  /**
   * 分片上传大文件（并行）
   */
  private async multipartUpload(objectKey: string, localPath: string, fileSize: number): Promise<void> {
    const partSize = this.PART_SIZE;
    const totalParts = Math.ceil(fileSize / partSize);
    const maxConcurrent = this.MAX_CONCURRENT_PARTS;
    
    logger.info(`📤 分片上传开始: ${totalParts} 个分片，每个分片 ${partSize / (1024 * 1024)}MB，并行数为 ${maxConcurrent}`);
    
    try {
      // 初始化分片上传
      const uploadId = await this.client.initMultipartUpload(objectKey);
      
      const parts: Array<{ number: number; etag: string }> = [];
      const startTime = Date.now();
      let uploadedBytes = 0;
      
      // 准备所有分片任务
      const uploadTasks = [];
      for (let i = 0; i < totalParts; i++) {
        const partNumber = i + 1;
        const start = i * partSize;
        const end = Math.min(start + partSize, fileSize);
        const partLength = end - start;
        
        uploadTasks.push({
          partNumber,
          start,
          partLength,
        });
      }
      
      // 并行上传分片
      const uploadPart = async (task: { partNumber: number; start: number; partLength: number }) => {
        const { partNumber, start, partLength } = task;
        
        // 读取分片数据
        const buffer = Buffer.allocUnsafe(partLength);
        const fileHandle = fs.openSync(localPath, 'r');
        fs.readSync(fileHandle, buffer, 0, partLength, start);
        fs.closeSync(fileHandle);
        
        const partStartTime = Date.now();
        logger.info(`📤 上传分片 ${partNumber}/${totalParts} - 大小: ${(partLength / (1024 * 1024)).toFixed(2)}MB`);
        
        try {
          const partResult = await this.client.uploadPart(
            objectKey,
            uploadId.uploadId,
            partNumber,
            buffer
          );
          
          const partElapsed = Date.now() - partStartTime;
          const speed = (partLength / (1024 * 1024)) / (partElapsed / 1000); // MB/s
          
          uploadedBytes += partLength;
          
          logger.info(`✅ 分片 ${partNumber} 上传完成 - 耗时: ${(partElapsed/1000).toFixed(1)}s, 平均速度: ${speed.toFixed(2)}MB/s`);
          
          return {
            number: partNumber,
            etag: partResult.etag
          };
        } catch (error) {
          logger.error(`❌ 分片 ${partNumber} 上传失败: ${error}`);
          throw error;
        }
      };
      
      // 使用Promise并发控制
      for (let i = 0; i < uploadTasks.length; i += maxConcurrent) {
        const batch = uploadTasks.slice(i, i + maxConcurrent);
        const batchResults = await Promise.all(batch.map(uploadPart));
        parts.push(...batchResults);
        
        const totalElapsed = Date.now() - startTime;
        const avgSpeed = uploadedBytes / (totalElapsed / 1000) / (1024 * 1024); // MB/s
        logger.info(`📊 批次完成 - 平均速度: ${avgSpeed.toFixed(2)}MB/s`);
      }
      
      // 按分片号排序
      parts.sort((a, b) => a.number - b.number);
      
      // 完成分片上传
      await this.client.completeMultipartUpload(objectKey, uploadId.uploadId, parts);
      
      const totalElapsed = Date.now() - startTime;
      const avgSpeed = fileSize / (1024 * 1024) / (totalElapsed / 1000);
      logger.info(`✅ 分片全部上传完成: ${totalParts} 个分片，总耗时: ${(totalElapsed/1000/60).toFixed(1)}分钟，平均速度: ${avgSpeed.toFixed(2)}MB/s`);
      
    } catch (error) {
      logger.error(`❌ 分片上传失败: ${error}`);
      throw error;
    }
  }
  
  /**
   * 生成预签名URL
   */
  async buildUrl(objectKey: string, expiresInHours: number = 24): Promise<string> {
    try {
      const expireSeconds = expiresInHours * 3600;
      
      // Generate presigned URL
      const signedUrl = this.client.signatureUrl(objectKey, {
        expires: expireSeconds,
        method: 'GET',
      });
      
      logger.info(`✅ 生成预签名URL: ${signedUrl}, 有效期: ${expiresInHours} 小时`);
      return signedUrl;
    } catch (error) {
      logger.error(`❌ 生成预签名URL失败: ${error}`);
      throw new Error(`生成预签名URL失败: ${error}`);
    }
  }
  
  /**
   * 解析阿里云OSS URL
   */
  parseUrl(url: string): { bucket: string; objectKey: string } {
    try {
      const parsed = new URL(url);
      const hostname = parsed.hostname || '';
      
      // Aliyun OSS URL format: https://bucket.oss-region.aliyuncs.com/path/file
      if (hostname.includes('.aliyuncs.com')) {
        const hostParts = hostname.split('.');
        if (hostParts.length >= 3) {
          const bucket = hostParts[0];
          const objectKey = decodeURIComponent(parsed.pathname.substring(1));
          return { bucket, objectKey };
        }
      }
      
      throw new Error(`无法解析阿里云OSS URL: ${url}`);
    } catch (error) {
      logger.error(`解析URL失败: ${error}`);
      throw new Error(`解析URL失败: ${error}`);
    }
  }
}
